crates/crabe_math/src/shape/line.rs

```diff
@@ -12,3 +12,148 @@ pub struct Line {
+        let line_direction = self.end - self.start;
+        let point_direction = *point - self.start;
+
+        let line_length_squared = line_direction.norm_squared();
+        if line_length_squared == 0.0 {
+            // The line segment has zero length, return the start point.
+            return self.start;
+        }
+        let t = point_direction.dot(&line_direction) / line_length_squared;
+
+        // The point is closest to a point on the segment.
+        self.start + t * line_direction
+    }
+
+    // return the closest point on the segment
+    pub fn closest_point_on_segment(&self, point: &Point2<f64>) -> Point2<f64> {
+        let line_direction = self.end - self.start;
+        let point_direction = *point - self.start;
+
+        let line_length_squared = line_direction.norm_squared();
+        if line_length_squared == 0.0 {
+            // The line segment has zero length, return the start point.
+            return self.start;
+        }
+
+        let t = point_direction.dot(&line_direction) / line_length_squared;
+        if t < 0.0 {
+            // The point is closest to the start of the segment.
+            return self.start;
+        } else if t > 1.0 {
+            // The point is closest to the end of the segment.
+            return self.end;
+        }
+
+        // The point is closest to a point on the segment.
+        self.start + t * line_direction
+    }
+
+    // return the closest point on the segment if it falls on him
+    pub fn orthogonal_projection_point_on_segment(
+        &self,
+        point: &Point2<f64>,
+    ) -> Option<Point2<f64>> {
+        let line_direction = self.end - self.start;
+        let point_direction = *point - self.start;
+
+        let line_length_squared = line_direction.norm_squared();
+        if line_length_squared == 0.0 {
+            // The line segment has zero length, return the start point.
+            return None;
+        }
+
+        let t = point_direction.dot(&line_direction) / line_length_squared;
+        if t < 0. || t > 1. {
+            return None;
+        } // The point don't fall on the segment.
+
+        // The point is closest to a point on the segment.
+        Some(self.start + t * line_direction)
+    }
+
+    // takes a number between 0 and 1 and return the point along the line situated at the specified dist
+    // for example with 0. it will return the starting point, and with 0.5 it will return the center of the line
+    pub fn point_allong_line(&self, x: f64) -> Point2<f64> {
+        let p_x = self.start.x + x * (self.end.x - self.start.x);
+        let p_y = self.start.y + x * (self.end.y - self.start.y);
+        return Point2::new(p_x, p_y);
+    }
+
+    // return the middle point of the segment
+    pub fn middle(&self) -> Point2<f64> {
+        return self.point_allong_line(0.5);
+    }
+
+    // return the length of the line
+    pub fn norm(&self) -> f64 {
```

Comment:
Can we use the norm function on the crate nalgebra for a vector instead to do our own.